### PR TITLE
Simple SpecFlow Examples

### PR DESCRIPTION
**Pull request by psiberknetic**
**Some code and examples for using SpecFlow in Visual Studio.**

_Note:_ This code was built as .NET Core 3.x. Building and running this code requires Visual Studio 2019 with .NET Core 3.x installed. Core 3.x can be installed via the Visual Studio Installer if it is not installed on your system.

Using SpecFlow also requires a one-time installation of the SpecFlow Visual Studio extension. This can be installed using the "Manage Extensions" option in Visual Studio.
image

To add a SpecFlow test to a test project, right click on the project and select Add, New Item and from the Add New Item screen select SpecFlow Feature File.
image

Once Gherkin has been written, generate the code behind by right-clicking within the Feature File and select "Generate Step Definitions"
image

### CODE REVIEW

**Comment by psiberknetic** [comment]
In the previous few commits, I've demonstrated some very simple SpecFlow scenarios that add numbers together.

This seems like a lot of work to implement multiple scenarios. Let's learn how to make this code simpler by using variables.

**Comment by psiberknetic** [comment]
Now that we've created scenarios with variables, we can continue to add new scenarios using the variable lines which create new tests without writing any additional code behind.

Additionally, testers writing Gherkin get intellisense style help as they write new scenarios.
image

Hooray for passing tests with VERY little code:
image

**Comment by psiberknetic** [comment]
You don't have to use Given, When and Then. The previously committed change demonstrations using just a Given and Then for a very simple test.

File: GettingStarted/3_TestToSeeIfOurStoreIsOpen.feature [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+	Given It is currently 12/15/2019 at 11:00PM
+	Then The store is closed
+
+Scenario: Open time edge case - store opens at 8AM
+	Given It is currently 12/11/2019 at 8:00AM
+	Then The store is open
+
+Scenario: Close time edge case - store closes at 10:00PM
+	Given It is currently 12/11/2019 at 10:00PM
+	Then The store is closed
```
Comment (psiberknetic) [inline GettingStarted/3_TestToSeeIfOurStoreIsOpen.feature:26]: You don't need to use Given, When and Then for every scenario. If you don't need one, don't use them.  Some tests are simple enough that you don't need to try to fit that template. 

In the "Check if the store is open" example above, just the Given and Then lines provide extremely readable test scenarios.

File: GettingStarted/IsStoreOpenScenarioOutlineVersion.feature [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+	Examples:
+		| date       | time     | isornot |
+		| 12/20/2019 | 12:00 PM | is      |
+		| 12/12/2019 | 9:00 AM  | is      |
+		| 01/03/2019 | 6:00 AM  | isnot   |
+		| 12/12/2019 | 11:00 PM | isnot   |
+		| 12/14/2019 | 9:00 AM  | isnot   |
+		| 12/14/2019 | 11:00 AM | is      |
+		| 12/12/2019 | 8:00 AM  | is      |
+		| 12/12/2019 | 10:00 PM | isnot   |
```
Comment (psiberknetic) [inline GettingStarted/IsStoreOpenScenarioOutlineVersion.feature:24]: Running many scenarios with the same logic can often be simplified by using Scenario Outlines. Scenario outlines allow you to provide a table of examples that will all run through the same check. Here is an example of the Gherkin text for testing whether our store is open at various dates and times.

One advantage of this practice is that new scenarios can be added to the table at any time, and the tests will be generated for them without having to change any code.

File: GettingStarted/CheckToSeeIfTheStoreIsOpen_OutlineVersionSteps.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+            if (dateToTest.IsStoreOpen())
+            {
+                p0.ToLower().Replace(" ", "").Should().Be("is");
+            }
+            else
+            {
+                p0.ToLower().Replace(" ", "").Should().Be("isnot");
+            }
+        }
+    }
```
Comment (psiberknetic) [inline GettingStarted/CheckToSeeIfTheStoreIsOpen_OutlineVersionSteps.cs:31]: And here is the simple code-behind that was implemented to run the scenarios.
